File: arithmetic_game.py
import time
import random
import sys
import json

REVIEW_LOG_FILE = 'review_log.txt'


def randomInt(min_val, max_val):
    """Return a random integer between min_val and max_val inclusive."""
    return random.randint(min_val, max_val)


def load_review_log():
    """Load persistent review questions from a file. Returns a list of review items."""
    try:
        with open(REVIEW_LOG_FILE, 'r') as f:
            data = json.load(f)
            return data
    except Exception:
        return []


def save_review_log(review_list):
    """Save persistent review questions to a file."""
    try:
        with open(REVIEW_LOG_FILE, 'w') as f:
            json.dump(review_list, f)
    except Exception as e:
        print(f"Error saving review log: {e}")


# Configuration: ask user for phase details
# Updated to allow toggling between specific and range in a single prompt

def configure_game():
    phases = []
    while True:
        try:
            num_phases = int(input("Enter number of phases (1-3): "))
            if 1 <= num_phases <= 3:
                break
            else:
                print("Please enter a number between 1 and 3.")
        except ValueError:
            print("Invalid input. Please enter an integer.")

    for i in range(1, num_phases + 1):
        print(f"\n--- Configuring Phase {i} ---")
        phase = {}
        # For first number: ask type
        while True:
            first_type = input(f"Phase {i}: Enter 's' for specific first number or 'r' for range: ").strip().lower()
            if first_type in ['s', 'r']:
                break
            else:
                print("Please enter 's' or 'r'.")
        if first_type == 's':
            while True:
                try:
                    value = int(input("Enter the specific first number (0-100): "))
                    if 0 <= value <= 100:
                        phase['firstSpecific'] = value
                        break
                    else:
                        print("Number must be between 0 and 100.")
                except ValueError:
                    print("Invalid input. Please enter an integer.")
        else:
            while True:
                try:
                    maximum = int(input("Enter the maximum for first number (0-100): "))
                    if 0 <= maximum <= 100:
                        phase['firstNum'] = maximum
                        break
                    else:
                        print("Number must be between 0 and 100.")
                except ValueError:
                    print("Invalid input. Please enter an integer.")

        # For second number
        while True:
            second_type = input(f"Phase {i}: Enter 's' for specific second number or 'r' for range: ").strip().lower()
            if second_type in ['s', 'r']:
                break
            else:
                print("Please enter 's' or 'r'.")
        if second_type == 's':
            while True:
                try:
                    value = int(input("Enter the specific second number (0-100): "))
                    if 0 <= value <= 100:
                        phase['secondSpecific'] = value
                        break
                    else:
                        print("Number must be between 0 and 100.")
                except ValueError:
                    print("Invalid input. Please enter an integer.")
        else:
            while True:
                try:
                    maximum = int(input("Enter the maximum for second number (0-100): "))
                    if 0 <= maximum <= 100:
                        phase['secondNum'] = maximum
                        break
                    else:
                        print("Number must be between 0 and 100.")
                except ValueError:
                    print("Invalid input. Please enter an integer.")

        # Operation
        while True:
            op = input("Enter the operation (+, -, *, /): ").strip()
            if op in ['+', '-', '*', '/']:
                phase['operation'] = op
                break
            else:
                print("Invalid operation. Please enter one of +, -, *, /.")

        # Limit type
        while True:
            print("Choose limit type:")
            print("1. Fixed number of questions")
            print("2. Time limit in seconds")
            choice = input("Enter 1 or 2: ").strip()
            if choice == '1':
                phase['limitType'] = 'questions'
                while True:
                    try:
                        q_limit = int(input("Enter the number of questions for this phase: "))
                        if q_limit > 0:
                            phase['limit'] = q_limit
                            break
                        else:
                            print("Must be greater than 0.")
                    except ValueError:
                        print("Invalid input. Please enter an integer.")
                break
            elif choice == '2':
                phase['limitType'] = 'time'
                while True:
                    try:
                        t_limit = int(input("Enter the time limit (in seconds) for this phase: "))
                        if t_limit > 0:
                            phase['limit'] = t_limit
                            break
                        else:
                            print("Must be greater than 0.")
                    except ValueError:
                        print("Invalid input. Please enter an integer.")
                break
            else:
                print("Please choose either 1 or 2.")

        # Allow negative results
        phase['allowNegative'] = input("Allow negative results? (y/n): ").strip().lower() == 'y'

        phases.append(phase)
    return phases


def generate_question(config):
    """Generate an arithmetic question based on phase configuration."""
    if 'firstSpecific' in config:
        num1 = config['firstSpecific']
    else:
        num1 = randomInt(0, config['firstNum'])
    
    if 'secondSpecific' in config:
        num2 = config['secondSpecific']
    else:
        # For division, avoid zero if possible
        if config['operation'] == '/' and config.get('secondNum', 0) == 0:
            num2 = 1
        else:
            num2 = randomInt(0 if config['operation'] != '/' else 1, config['secondNum'])

    op = config['operation']
    if op == '+':
        answer = num1 + num2
        question = f"What is {num1} + {num2}?"
    elif op == '-':
        # Ensure non-negative if not allowed
        if not config['allowNegative'] and num1 < num2:
            num1, num2 = num2, num1
        answer = num1 - num2
        question = f"What is {num1} - {num2}?"
    elif op == '*':
        answer = num1 * num2
        question = f"What is {num1} * {num2}?"
    elif op == '/':
        if num2 == 0:
            num2 = 1
        if 'firstSpecific' in config:
            num1 = num1 - (num1 % num2)
        else:
            max_quotient = config.get('firstNum', 100) // num2
            quotient = randomInt(0, max_quotient) if max_quotient > 0 else 0
            num1 = quotient * num2
        answer = num1 // num2
        question = f"What is {num1} / {num2}? (Provide the integer quotient)"
    else:
        answer = num1 + num2
        question = f"What is {num1} + {num2}?"
    return question, answer


def ask_question(prompt, correct_answer):
    """Prompt the user until they get the answer correct, counting wrong attempts."""
    wrong_attempts = 0
    while True:
        user_input = input(prompt + "\nYour answer: ")
        try:
            user_ans = int(user_input.strip())
        except ValueError:
            print("Please enter a valid integer answer.")
            continue
        if user_ans == correct_answer:
            print(f"Correct! You answered: {user_ans}\n")
            return wrong_attempts
        else:
            wrong_attempts += 1
            print(f"Incorrect, you answered: {user_ans}. This problem will be reviewed later.\n")


def run_phase(phase_config, phase_number, overall_start):
    print("\n\n--- Phase {} ---\n".format(phase_number))
    review_queue = []
    # For questions mode
    if phase_config['limitType'] == 'questions':
        total = phase_config['limit']
        for q in range(1, total + 1):
            overall_elapsed = int(time.time() - overall_start)
            question_prompt = f"({overall_elapsed} sec elapsed) Question {q}:"
            question_text, answer = generate_question(phase_config)
            # Display page break is already done by printing newlines above
            # Bold formatting using ANSI escape codes if supported
            print(f"\033[1m{question_text}\033[0m")
            wrong = ask_question(question_prompt, answer)
            if wrong > 0:
                # Add to review queue with required repetitions equal to wrong attempts
                review_queue.append({'question': question_text, 'answer': answer, 'required': wrong, 'correctCount': 0})
    else:
        # Time mode
        limit = phase_config['limit']
        q = 0
        phase_start = time.time()
        while True:
            elapsed = int(time.time() - phase_start)
            if elapsed >= limit:
                break
            q += 1
            overall_elapsed = int(time.time() - overall_start)
            question_prompt = f"({overall_elapsed} sec elapsed) Question {q}:"
            question_text, answer = generate_question(phase_config)
            print(f"\033[1m{question_text}\033[0m")
            wrong = ask_question(question_prompt, answer)
            if wrong > 0:
                review_queue.append({'question': question_text, 'answer': answer, 'required': wrong, 'correctCount': 0})
    
    # Process review questions until all are mastered
    if review_queue:
        print("\nReview Round for Phase {}:\n".format(phase_number))
    while review_queue:
        for item in review_queue[:]:
            overall_elapsed = int(time.time() - overall_start)
            prompt = f"({overall_elapsed} sec elapsed) Review:"
            print(f"\033[1m{item['question']}\033[0m")
            wrong = ask_question(prompt, item['answer'])
            # If answered correctly, increment correct count
            item['correctCount'] += 1
            if item['correctCount'] >= item['required']:
                review_queue.remove(item)
    print(f"--- Completed Phase {phase_number} ---\n")


def end_menu(overall_start):
    total_time = int(time.time() - overall_start)
    print(f"Congratulations! You completed the game in {total_time} seconds.")
    while True:
        choice = input("Enter 'r' to restart or 'q' to quit: ").strip().lower()
        if choice == 'r':
            main()
            break
        elif choice == 'q':
            print("Goodbye!")
            # Before quitting, save persistent review log
            persistent_reviews = load_review_log()
            # Append any remaining review questions (this simplistic approach: just append current session reviews)
            # In this implementation, review questions have already been mastered, so we clear the log
            save_review_log([])
            sys.exit(0)
        else:
            print("Invalid input. Please enter 'r' or 'q'.")


def main():
    print("Welcome to the Arithmetic Game!\n")
    phases_config = configure_game()
    # Load persistent review questions from previous sessions
    persistent_reviews = load_review_log()
    overall_start = time.time()
    phase_number = 1
    for config in phases_config:
        run_phase(config, phase_number, overall_start)
        phase_number += 1
    # After main phases, include persistent review questions if any
    if persistent_reviews:
        print("\nReview Round for Persistent Problems:\n")
        while persistent_reviews:
            for item in persistent_reviews[:]:
                overall_elapsed = int(time.time() - overall_start)
                prompt = f"({overall_elapsed} sec elapsed) Persistent Review:"
                print(f"\033[1m{item['question']}\033[0m")
                wrong = ask_question(prompt, item['answer'])
                item['correctCount'] += 1
                if item['correctCount'] >= item['required']:
                    persistent_reviews.remove(item)
        save_review_log(persistent_reviews)
    end_menu(overall_start)


if __name__ == "__main__":
    main()

File: arithmetic_game.js
/*
 * Arithmetic Game JavaScript implementation (Web Version with Configurable Phases and Review Mechanism)
 * This file provides a web-based interface for the arithmetic game where the user can configure up to 3 phases.
 * Each phase's configuration includes:
 * 1. First number maximum (or a specific number if chosen)
 * 2. Operation (+, -, *, /)
 * 3. Second number maximum (or a specific number if chosen)
 * 4. Limit type: either a fixed number of questions or a time limit (in seconds) for the phase
 * 5. Option to allow negative results
 * 
 * In addition, when a problem is answered incorrectly, the problem is tracked and repeated later in the problem set.
 * If a problem is answered incorrectly multiple times, it will be repeated until it has been answered correctly as many times as it was answered incorrectly.
 * Persistently, unresolved problems are stored and will reappear in future instances until mastered.
 * 
 * NEW ENHANCEMENTS:
 * - The arithmetic question is formatted with the operation placed between the first and second numbers.
 * - A log prevents any question from being repeated more than 3 times during the session.
 * - The same question will not be displayed consecutively.
 * - The question display format is updated to show:
 *       Phase 1
 *       1. What is 5 - 5?
 * - The end game message now displays the total questions, total time, and average time per question.
 */

// Utility function to generate a random integer between min and max (inclusive)
function randomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

// Globals for game state
let gamePhases = [];      // Array of phase objects { config, total }
let currentPhaseIndex = 0;
let currentQuestionIndex = 0;
let currentQuestion = null; // currentQuestion object will include { question, answer, fromReview (bool), reviewData (object) }
let startTime = null;     // Overall game start time
let phaseStartTime = null; // Start time for current phase
let timerInterval = null;

// Global tracking for question frequency and ordering
let totalQuestionsAnswered = 0;             // Total number of questions answered in this session
let questionLog = {};                       // Log mapping question text to the number of times it has been shown
let lastQuestion = "";                      // Last question text displayed

// Review queue for questions answered incorrectly
// Each item: { question: <string>, answer: <number>, required: <number>, correctCount: <number> }
let reviewQueue = [];

// LocalStorage key for persistent review data
const REVIEW_STORAGE_KEY = 'reviewQueue';

// UI Elements
const timerElem = document.getElementById('timer');
const questionElem = document.getElementById('question');
const answerInput = document.getElementById('answer');
const submitButton = document.getElementById('submit');
const feedbackElem = document.getElementById('feedback');
const endgamePanel = document.getElementById('endgame-panel');
const endgameMessage = document.getElementById('endgame-message');

// Persistent reviewQueue functions
function loadPersistentReviewQueue() {
    const data = localStorage.getItem(REVIEW_STORAGE_KEY);
    return data ? JSON.parse(data) : [];
}

function savePersistentReviewQueue() {
    localStorage.setItem(REVIEW_STORAGE_KEY, JSON.stringify(reviewQueue));
}

// Update the timer display (running timer displayed at top)
function updateTimer() {
    const elapsed = Math.floor((Date.now() - startTime) / 1000);
    timerElem.textContent = `${elapsed} seconds`;
}

// Generate a new question based on provided phase configuration
function generateQuestion(config) {
    let num1, num2, questionText, answer;
    switch (config.operation) {
        case '+':
            num1 = (config.firstSpecific !== undefined) ? config.firstSpecific : randomInt(0, config.firstNum);
            num2 = (config.secondSpecific !== undefined) ? config.secondSpecific : randomInt(0, config.secondNum);
            answer = num1 + num2;
            break;
        case '-':
            num1 = (config.firstSpecific !== undefined) ? config.firstSpecific : randomInt(0, config.firstNum);
            num2 = (config.secondSpecific !== undefined) ? config.secondSpecific : randomInt(0, config.secondNum);
            // Ensure non-negative if not allowed
            if (!config.allowNegative && num1 < num2) {
                if (config.firstSpecific !== undefined && config.secondSpecific !== undefined) {
                    [num1, num2] = [num2, num1];
                } else {
                    num1 = Math.max(num1, num2);
                }
            }
            answer = num1 - num2;
            break;
        case '*':
            num1 = (config.firstSpecific !== undefined) ? config.firstSpecific : randomInt(0, config.firstNum);
            num2 = (config.secondSpecific !== undefined) ? config.secondSpecific : randomInt(0, config.secondNum);
            answer = num1 * num2;
            break;
        case '/':
            if (config.secondSpecific !== undefined) {
                num2 = (config.secondSpecific === 0) ? 1 : config.secondSpecific;
            } else {
                num2 = randomInt(1, config.secondNum > 0 ? config.secondNum : 1);
            }
            if (config.firstSpecific !== undefined) {
                num1 = config.firstSpecific;
                num1 = num1 - (num1 % num2);
            } else {
                let maxQuotient = Math.floor(config.firstNum / num2);
                maxQuotient = maxQuotient < 0 ? 0 : maxQuotient;
                const quotient = randomInt(0, maxQuotient);
                num1 = quotient * num2;
            }
            answer = num1 / num2;
            break;
        default:
            num1 = (config.firstSpecific !== undefined) ? config.firstSpecific : randomInt(0, config.firstNum);
            num2 = (config.secondSpecific !== undefined) ? config.secondSpecific : randomInt(0, config.secondNum);
            answer = num1 + num2;
    }
    // Format the question with the operator between the numbers
    questionText = `What is ${num1} ${config.operation} ${num2}?`;
    return { question: questionText, answer: answer };
}

// Load the next question for the current phase
function loadNextQuestion() {
    // If all phases are completed, end game
    if (currentPhaseIndex >= gamePhases.length) {
        endGame();
        return;
    }
    const phase = gamePhases[currentPhaseIndex];
    const config = phase.config;

    // Check limit for phase
    if (config.limitType === 'time') {
        const phaseElapsed = Math.floor((Date.now() - phaseStartTime) / 1000);
        if (phaseElapsed >= config.limit) {
            currentPhaseIndex++;
            currentQuestionIndex = 0;
            phaseStartTime = Date.now();
            loadNextQuestion();
            return;
        }
    } else { // questions mode
        if (currentQuestionIndex >= phase.total) {
            currentPhaseIndex++;
            currentQuestionIndex = 0;
            phaseStartTime = Date.now();
            loadNextQuestion();
            return;
        }
    }

    let candidate = null;
    // Every 3 questions, if reviewQueue is non-empty, serve a review question
    if (reviewQueue.length > 0 && (currentQuestionIndex % 3 === 0)) {
        candidate = reviewQueue.shift();
        candidate.fromReview = true;
        // If this candidate has been displayed too many times or is the same as the last question, skip it
        if ((questionLog[candidate.question] || 0) >= 3 || candidate.question === lastQuestion) {
            loadNextQuestion();
            return;
        }
    } else {
        candidate = generateQuestion(config);
        candidate.fromReview = false;
        let attempts = 0;
        // Regenerate if candidate has been shown too many times or is same as last question
        while (((questionLog[candidate.question] || 0) >= 3 || candidate.question === lastQuestion) && attempts < 10) {
            candidate = generateQuestion(config);
            candidate.fromReview = false;
            attempts++;
        }
    }

    // Update logs
    questionLog[candidate.question] = (questionLog[candidate.question] || 0) + 1;
    lastQuestion = candidate.question;
    totalQuestionsAnswered++;

    currentQuestion = candidate;
    currentQuestionIndex++;
    // Update question display in the specified format:
    // Phase X
    // Y. <strong>Question</strong>
    questionElem.innerHTML = `Phase ${currentPhaseIndex + 1}<br>${currentQuestionIndex}. <strong>${currentQuestion.question}</strong>`;
    answerInput.value = '';
    feedbackElem.textContent = '';
}

// Process answer submission
function handleSubmit() {
    const userAnswer = parseInt(answerInput.value.trim(), 10);
    if (isNaN(userAnswer)) {
        feedbackElem.style.color = 'red';
        feedbackElem.textContent = 'Please enter a valid integer answer.';
        return;
    }
    if (userAnswer === currentQuestion.answer) {
        feedbackElem.style.color = 'green';
        feedbackElem.textContent = `Correct! You answered: ${userAnswer}.`;
        // If this was a review question, update its record
        if (currentQuestion.fromReview) {
            if (!currentQuestion.reviewData) {
                // Initialize review data
                currentQuestion.reviewData = { required: 1, correctCount: 0 };
            }
            currentQuestion.reviewData.correctCount++;
            if (currentQuestion.reviewData.correctCount < currentQuestion.reviewData.required) {
                // Not mastered yet, requeue the review question
                reviewQueue.push(currentQuestion);
            }
        }
        setTimeout(() => {
            savePersistentReviewQueue();
            loadNextQuestion();
        }, 1000);
    } else {
        feedbackElem.style.color = 'red';
        feedbackElem.textContent = `Incorrect, you answered: ${userAnswer}. This problem will be reviewed later.`;
        // Log this incorrect attempt in the review queue
        let found = false;
        // Check if this problem is already in reviewQueue (by question text)
        for (let item of reviewQueue) {
            if (item.question === currentQuestion.question) {
                item.required++;
                found = true;
                break;
            }
        }
        if (!found) {
            // If current question is already a review question, update its data; else create new review record
            if (currentQuestion.fromReview) {
                if (!currentQuestion.reviewData) {
                    currentQuestion.reviewData = { required: 1, correctCount: 0 };
                } else {
                    currentQuestion.reviewData.required++;
                }
                reviewQueue.push(currentQuestion);
            } else {
                // Create a new review record from the current question
                let reviewItem = {
                    question: currentQuestion.question,
                    answer: currentQuestion.answer,
                    required: 1,
                    correctCount: 0,
                    fromReview: true
                };
                reviewQueue.push(reviewItem);
            }
        }
        setTimeout(() => {
            savePersistentReviewQueue();
            loadNextQuestion();
        }, 1000);
    }
}

// End the game: hide answer input/submit button and show endgame panel
function endGame() {
    clearInterval(timerInterval);
    const totalTime = Math.floor((Date.now() - startTime) / 1000);
    answerInput.style.display = 'none';
    submitButton.style.display = 'none';
    const avgTime = totalQuestionsAnswered > 0 ? (totalTime / totalQuestionsAnswered).toFixed(2) : 0;
    endgameMessage.textContent = `You completed ${totalQuestionsAnswered} questions in ${totalTime} seconds (average ${avgTime} seconds per question).`;
    endgamePanel.style.display = 'block';
    savePersistentReviewQueue();
}

// Start the game using configuration from window.gameConfig
function startGame() {
    if (!window.gameConfig || !Array.isArray(window.gameConfig) || window.gameConfig.length === 0) {
        console.error('No game configuration provided.');
        return;
    }
    // Reset global logs for a new game session
    totalQuestionsAnswered = 0;
    questionLog = {};
    lastQuestion = "";

    reviewQueue = loadPersistentReviewQueue();
    gamePhases = window.gameConfig.map(config => {
        return { config: config, total: config.limit };
    });
    currentPhaseIndex = 0;
    currentQuestionIndex = 0;
    answerInput.style.display = 'inline-block';
    submitButton.style.display = 'inline-block';
    endgamePanel.style.display = 'none';
    startTime = Date.now();
    phaseStartTime = Date.now();
    timerInterval = setInterval(updateTimer, 1000);
    loadNextQuestion();
}

// Event listeners for submit
submitButton.addEventListener('click', handleSubmit);
answerInput.addEventListener('keydown', function(e) {
    if (e.key === 'Enter') {
        handleSubmit();
    }
});

updateTimer();
window.startGame = startGame;
